File: zig/src/object.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

const ast = @import("ast.zig");
const env = @import("environment.zig");

pub const ObjectTag = enum {
    integer,
    boolean,
    return_obj,
    function_obj,
    null_obj,
};

pub const Object = union(ObjectTag) {
    const Self = @This();

    integer: i64,
    boolean: bool,
    return_obj: *Object,
    function_obj: *FunctionObject,
    null_obj: void,
};

pub const FunctionObject = struct {
    const Self = @This();

    allocator: Allocator,

    parameters: std.ArrayList([]const u8),
    body: ast.BlockStatement,
    environment: *env.Environment,

    pub fn init(allocator: Allocator, parameters: [][]const u8, body: ast.BlockStatement, environment: *env.Environment) Allocator.Error!Self {
        var params = try std.ArrayList([]const u8).initCapacity(allocator, parameters.len);
        errdefer {
            for (params.items) |param| {
                allocator.free(param);
            }
            params.deinit();
        }

        for (parameters) |param| {
            const param_copy = try allocator.alloc(u8, param.len);
            errdefer allocator.free(param_copy);

            @memcpy(param_copy, param);

            try params.append(param_copy);
        }

        return Self{
            .allocator = allocator,
            .parameters = params,
            .body = try body.clone(),
            .environment = environment,
        };
    }

    pub fn deinit(self: Self) void {
        for (self.parameters.items) |param| {
            self.allocator.free(param);
        }
        self.parameters.deinit();

        self.body.deinit();
    }
};

File: rust/src/object.rs
use crate::{ast, environment::Environment};

use std::{collections::HashMap, fmt::Display, rc::Rc};

#[derive(Debug, PartialEq, Clone)]
pub enum Object {
    Integer(i64),
    String(Rc<String>),
    Boolean(bool),
    Return(Rc<Object>),
    Function(FunctionObject),
    Builtin(BuiltinFunction),
    Array(Rc<Vec<Object>>),
    HashMap(Rc<HashMap<HashKey, Object>>),
    Null,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum DataType {
    Integer,
    String,
    Boolean,
    Return,
    Function,
    Builtin,
    Array,
    HashMap,
    Null,
}

impl From<&Object> for DataType {
    fn from(value: &Object) -> Self {
        match value {
            Object::Integer(_) => Self::Integer,
            Object::String(_) => Self::String,
            Object::Boolean(_) => Self::Boolean,
            Object::Return(_) => Self::Return,
            Object::Function(_) => Self::Function,
            Object::Builtin(_) => Self::Builtin,
            Object::Array(_) => Self::Array,
            Object::HashMap(_) => Self::HashMap,
            Object::Null => Self::Null,
        }
    }
}

impl From<Object> for DataType {
    fn from(value: Object) -> Self {
        (&value).into()
    }
}

impl Display for DataType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let string = match self {
            DataType::Integer => "INTEGER",
            DataType::String => "STRING",
            DataType::Boolean => "BOOLEAN",
            DataType::Return => "RETURN",
            DataType::Function => "FUNCTION",
            DataType::Builtin => "BUILTIN",
            DataType::Array => "ARRAY",
            DataType::HashMap => "HASH_MAP",
            DataType::Null => "NULL",
        };

        f.write_str(string)
    }
}

impl Object {
    pub fn inspect(&self) -> String {
        match self {
            Object::Integer(i) => i.to_string(),
            Object::String(s) => (**s).clone(),
            Object::Boolean(b) => b.to_string(),
            Object::Return(o) => o.inspect(),
            Object::Function(fun) => fun.inspect(),
            Object::Builtin(f) => format!("builtin function {:?}", f),
            Object::Array(arr) => {
                let elements = arr
                    .iter()
                    .map(|obj| obj.inspect())
                    .collect::<Vec<_>>()
                    .join(", ");
                format!("[{}]", elements)
            }
            Object::HashMap(map) => {
                let elements = map
                    .iter()
                    .map(|(key, value)| format!("{}: {}", key.inspect(), value.inspect()))
                    .collect::<Vec<_>>()
                    .join(", ");

                format!("{{{}}}", elements)
            }
            Object::Null => "null".to_string(),
        }
    }

    pub fn is_truthy(&self) -> bool {
        !matches!(self, Object::Boolean(false) | Object::Null)
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct FunctionObject {
    pub parameters: Rc<Vec<String>>,
    pub body: ast::BlockStatement,
    pub environment: Environment,
}

impl FunctionObject {
    fn inspect(&self) -> String {
        format!(
            "fn ({}) {{\n{}\n}}",
            self.parameters.join(", "),
            self.body.debug_str(),
        )
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum HashKey {
    String(Rc<String>),
    Integer(i64),
    Boolean(bool),
}

impl HashKey {
    fn inspect(&self) -> String {
        match self {
            Self::String(s) => (**s).clone(),
            Self::Integer(i) => i.to_string(),
            Self::Boolean(b) => b.to_string(),
        }
    }
}

impl TryFrom<Object> for HashKey {
    type Error = DataType;

    fn try_from(value: Object) -> Result<Self, Self::Error> {
        match value {
            Object::String(str) => Ok(Self::String(str)),
            Object::Integer(i) => Ok(Self::Integer(i)),
            Object::Boolean(b) => Ok(Self::Boolean(b)),
            _ => Err(value.into()),
        }
    }
}

#[derive(Debug, PartialEq, Clone)]
pub enum BuiltinFunction {
    Len,
    First,
    Last,
    Rest,
    Push,
    Puts,
}

impl BuiltinFunction {
    pub(crate) fn from_ident(ident: &str) -> Option<Self> {
        match ident {
            "len" => Some(Self::Len),
            "first" => Some(Self::First),
            "last" => Some(Self::Last),
            "rest" => Some(Self::Rest),
            "push" => Some(Self::Push),
            "puts" => Some(Self::Puts),
            _ => None,
        }
    }
}
